File: synths/korg-dw8000/KorgDW8000.cpp
/*
   Copyright (c) 2020 Christof Ruch. All rights reserved.

   Dual licensed: Distributed under Affero GPL license by default, an MIT license is available for purchase
*/

#include "KorgDW8000.h"

#include "KorgDW8000Patch.h"
#include "KorgDW8000Parameter.h"
//#include "KorgDW8000_BCR2000.h"

//#include "BCR2000.h"

#include "MidiHelpers.h"
#include "Sysex.h"

#include <iterator>
#include <boost/format.hpp>

namespace midikraft {

	std::string KorgDW8000::getName() const
	{
		return "Korg DW 8000";
	}

	juce::MidiMessage KorgDW8000::deviceDetect(int channel)
	{
		// Build Device ID request message
		return MidiHelpers::sysexMessage({ 0x42 /* Korg */, uint8(0x40 | channel) });
	}

	int KorgDW8000::deviceDetectSleepMS()
	{
		// The Korg is reasonably fast to reply, try 100 ms
		return 100;
	}

	MidiChannel KorgDW8000::channelIfValidDeviceResponse(const MidiMessage &message) {
		// Is this the correct Device ID message?
		auto data = message.getSysExData();
		if (message.isSysEx() && data[0] == 0x42 /* Korg */
			&& (data[1] & 0xf0) == 0x30  /* Device ID */
			&& data[2] == 0x03 /* DW 8000 */) {
			return MidiChannel::fromZeroBase(data[1] & 0x0f);
		}
		return MidiChannel::invalidChannel();
	}

	bool KorgDW8000::needsChannelSpecificDetection()
	{
		return true;
	}

	bool KorgDW8000::isOwnSysex(MidiMessage const &message) const
	{
		auto data = message.getSysExData();
		return message.isSysEx() && data[0] == 0x42 /* Korg */
			&& ((data[1] & 0xf0) == 0x30) /* Format */
			//&& ((data[1] & 0x0f) == midiChannel_) /* correct channel */
			&& data[2] == 0x03; /* DW 8000 */
	}

	int KorgDW8000::numberOfBanks() const
	{
		return 1;
	}

	juce::MidiMessage KorgDW8000::requestEditBufferDump()
	{
		// This is called a "Data Save Request" in the Service Manual (p. 6)
		return MidiHelpers::sysexMessage({ 0x42 /* Korg */, uint8(0x30 | channel().toZeroBasedInt()), 0x03 /* Model ID = DW 8000 */, DATA_SAVE_REQUEST });
	}

	/*
	std::vector<juce::MidiMessage> KorgDW8000::requestPatch(int patchNo)
	{
		std::vector<MidiMessage> result;

		// We need to send a program change
		result.push_back(MidiMessage::programChange(channel().toOneBasedInt(), patchNo));
		// And request the edit buffer
		result.push_back(requestEditBufferDump());
		return result;
	}*/

	int KorgDW8000::numberOfPatches() const
	{
		return 64;
	}

	std::string KorgDW8000::friendlyBankName(MidiBankNumber bankNo) const
	{
		ignoreUnused(bankNo);
		return "Standard Bank";
	}

	juce::MidiMessage KorgDW8000::saveEditBufferToProgram(int programNumber)
	{
		// The Korg DW 8000 has no direct download to program slot message - you can only send it the edit buffer, and then instruct it to 
		// save it into a program you specify
		jassert(programNumber >= 0 && programNumber < 64);
		if (programNumber < 0 || programNumber >= 64) {
			return MidiMessage();
		}
		return MidiHelpers::sysexMessage({ 0x42 /* Korg */, uint8(0x30 | channel().toZeroBasedInt()),
			0x03 /* Model ID = DW 8000 */, WRITE_REQUEST, uint8(programNumber) });
	}

	MidiChannel KorgDW8000::getInputChannel() const
	{
		return channel();
	}

	/*std::vector<std::string> KorgDW8000::presetNames()
	{
		return { (boost::format("Knobkraft %s") % getName()).str() };
	}*/

	bool KorgDW8000::isEditBufferDump(const MidiMessage& message) const
	{
		auto data = message.getSysExData();
		return isOwnSysex(message)
			&& data[3] == DATA_DUMP;
	}

	std::shared_ptr<Patch> KorgDW8000::patchFromSysex(const MidiMessage& message) const
	{
		// The DW8000 is so primitive that it does do nothing to the few bytes of data it needs per patch
		if (!isEditBufferDump(message)) {
			jassert(false);
			return {};
		}

		// Extract the data
		auto data = message.getSysExData();
		Synth::PatchData patchdata;
		for (int index = 4; index < message.getSysExDataSize(); index++) {
			patchdata.push_back(data[index]);
		}
		return {std::make_shared<KorgDW8000Patch>(patchdata, MidiProgramNumber::fromZeroBase(0))};
	}

	std::shared_ptr<DataFile> KorgDW8000::patchFromPatchData(const Synth::PatchData &data, MidiProgramNumber place) const {
		return std::make_shared<KorgDW8000Patch>(data, place);
	}

	std::vector<juce::MidiMessage> KorgDW8000::patchToSysex(const Patch &patch) const
	{
		// For the patch sysex, all we need to do is to add the header on the correct channel
		std::vector<uint8> data(4);
		data[0] = 0x42; // Korg
		data[1] = uint8(0x30 | channel().toZeroBasedInt());
		data[2] = 0x03; // DW8000
		data[3] = DATA_DUMP;
		std::copy(patch.data().begin(), patch.data().end(), std::back_inserter(data));

		return std::vector<MidiMessage>({ MidiHelpers::sysexMessage(data) });
	}

	/*using namespace BinaryData;

	// These are the filenames for the various ROMs from the Internet which contain the real wave forms of the DWGS system
	// http://dbwbp.com/index.php/9-misc/37-synth-eprom-dumps
	std::vector<const char *> kDR8000ROMs = {
		HN613256PT70_bin, HN613256PT71_bin, HN613256PCB4_bin, HN613256PCB5_bin, EXP1_bin, EXP3_bin, EXP2_bin, EXP4_bin
	};

	std::vector<float> KorgDW8000::romWave(int waveNo)
	{
		std::vector<float> result;
		if (waveNo < 0 || waveNo > 32) {
			jassert(false);
			return result;
		}

		// Select the right ROM binary resource
		int romNo = waveNo / 4;
		const uint8 *rom = reinterpret_cast<const uint8 *>(kDR8000ROMs[romNo]);

		// Each ROM contains 4 waves, 8bit unsigned format, with the following data:
		//2048 samples octave 0
		//2048 samples octave 1
		//1024 samples octave 2
		//1024 samples octave 3
		//512 samples octave 4
		//512 samples octave 5
		//512 samples octave 6
		//512 samples octave 7

		int base = (waveNo % 4) * 8192; // 8k per wave, 4 waves per file. The Kawai K3 is much more subtle in its design
		for (int step = 0; step < 2048; step++) {
			result.push_back(rom[base + step]);
		}
		return result;
	}

	std::vector<std::string> waveNames = {
		"Sawtooth", "Square", "Piano", "Electric piano 1", "Electric piano 2", "Clavinet", "Organ", "Brass", "Sax", "Violin", "Guitar", "Electric guitar", "Bass", "Digital bass", "Bell and whistle"
	};

	std::string KorgDW8000::waveName(int waveNo)
	{
		if (waveNo < 0 || waveNo > 15) {
			jassert(false);
			return "unknown";
		}
		return waveNames[waveNo];
	}

	void KorgDW8000::createReverseEngineeringData() {
		std::vector<juce::MidiMessage> reverseEngineer;


		// Ok, 51 parameters to fit into 64 patches... that should be easy.
		for (int patch = 0; patch < 64; patch++) {
			// Create completely empty patch - 51 bytes 0
			std::vector<uint8> data(51);

			if (patch < KorgDW8000Parameter::allParameters.size()) {
				// set this one parameter to its max value
				data[KorgDW8000Parameter::allParameters[patch]->sysexIndex()] = (uint8)KorgDW8000Parameter::allParameters[patch]->maxValue();
			}

			KorgDW8000Patch testPatch(data, (boost::format("Patch %d") % patch).str());
			reverseEngineer.push_back(patchToSysex(testPatch)[0]);
			reverseEngineer.push_back(saveEditBufferToProgram(patch));
		}

		Sysex::saveSysex("reveng.syx", reverseEngineer);
	}

	void KorgDW8000::setupBCR2000(MidiController *controller, BCR2000 &bcr, SimpleLogger *logger) {
		if (!bcr.channel().isValid()) return;
		if (!channel().isValid()) return;

		auto bcl = KorgDW8000BCR2000::generateBCL(channel().toZeroBasedInt());
		auto syx = bcr.convertToSyx(bcl);
		controller->enableMidiInput(bcr.midiInput()); // Make sure we listen to the answers from the BCR2000 that we detected!
		bcr.sendSysExToBCR(controller->getMidiOutput(bcr.midiOutput()), syx, *controller, logger);
	}

	void KorgDW8000::syncDumpToBCR(MidiProgramNumber programNumber, MidiController *controller, BCR2000 &bcr, SimpleLogger *logger) {
	}

	void KorgDW8000::setupBCR2000View(BCR2000_Component &view) {
		//TODO
	}

	void KorgDW8000::setupBCR2000Values(BCR2000_Component &view, Patch *patch)
	{
		//TODO
	}*/

}

File: adaptions/KorgDW8000.py
#
#   Copyright (c) 2023 Christof Ruch. All rights reserved.
#
#   Dual licensed: Distributed under Affero GPL license by default, an MIT license is available for purchase
#
from typing import List

import testing
from orm_synth import *


class Korg_DW8000(Synth, DiscoverableDevice, EditBufferCapability, BankDescriptorsCapability):

    def __init__(self):
        Synth.__init__(self)
        DiscoverableDevice.__init__(self)
        EditBufferCapability.__init__(self)
        BankDescriptorsCapability.__init__(self)
        self.channel = 0

    def get_name(self):
        return "Korg DW-8000 Adaptation"

    def bank_descriptors(self) -> List[BankDescriptor]:
        return [BankDescriptor(bank=Bank(0, 64), name="Internal", size=64, type="Patch", is_rom=False)]

    def friendly_program_name(self, bank: Bank, program: Program) -> str:
        return "%d%d" % (program.value() // 8 + 1, (program.value() % 8) + 1)

    def needs_channel_specific_detection(self) -> bool:
        return True

    def create_device_detect_message(self, channel) -> List[MidiMessage]:
        # Page 5 of the service manual - Device ID Request
        return [MidiMessage([0xf0, 0x42, 0x40 | (channel % 16), 0xf7])]

    def channel_if_valid_device_detect_response(self, message: MidiMessage) -> int:
        # Page 3 of the service manual - Device ID
        if (len(message) > 3
                and message[0] == 0xf0  # Sysex
                and message[1] == 0x42  # Korg
                and (message[2] & 0xf0) == 0x30  # Device ID
                and message[3] == 0x03):  # DW-8000
            self.channel = message[2] & 0x0f
            return self.channel
        return -1

    def device_detect_sleep_ms(self) -> int:
        return 100

    def request_edit_buffer(self) -> List[MidiMessage]:
        # Page 5 - Data Save Request
        return [MidiMessage([0xf0, 0x42, 0x30 | self.channel, 0x03, 0x10, 0xf7])]

    def is_edit_buffer(self, messages: List[MidiMessage]) -> bool:
        # Page 3 - Data Dump
        message = messages[0]
        return (len(message) > 4
                and message[0] == 0xf0
                and message[1] == 0x42  # Korg
                and (message[2] & 0xf0) == 0x30  # Format, ignore MIDI Channel in lower 4 bits
                and message[3] == 0x03  # DW-8000
                and message[4] == 0x40  # Data Dump
                )

    def convert_to_edit_buffer(self, patch: List[MidiMessage]) -> List[MidiMessage]:
        if self.is_edit_buffer(patch):
            return [MidiMessage(patch[0][0:2] + [0x30 | self.channel] + patch[0][3:])]
        raise Exception("This is not an edit buffer - can't be converted")


def make_test_data():
    def programs(data: testing.TestData) -> List[testing.ProgramTestData]:
        patch = [x for sublist in data.all_messages[0:2] for x in sublist]
        yield testing.ProgramTestData(message=patch)

    # assert friendlyProgramName(0) == "11"
    return testing.TestData(sysex="testData/KorgDW8000_bank_a.syx", edit_buffer_generator=programs, friendly_bank_name=(63, "88"))


if __name__ == "__main__":
    korg = Korg_DW8000()
    print(korg.get_name())
    print(korg.friendly_program_name(Bank(0, 64), Program.from_zero_base(17)))
    print(korg.create_device_detect_message(3))
    print(korg.bank_descriptors()[0].name)
